### PR TITLE
Performance improvement for subquery with MySQL  

### PR DESCRIPTION
**Pull request by jewome62**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        |

Add orderBy into subrequest
Remove orderBy from main request if added into subrequest
Do OrderExtension before FilterEagerLoadingExtension

### CODE REVIEW

File: src/Bridge/Symfony/Bundle/Resources/config/doctrine_orm.xml [hunk 1]
```diff
@@ -135,6 +135,14 @@
         </service>
         <service id="ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\EagerLoadingExtension" alias="api_platform.doctrine.orm.query_extension.eager_loading" />
 
+        <service id="api_platform.doctrine.orm.query_extension.order" class="ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\OrderExtension" public="false">
```
Comment (jewome62) [inline src/Bridge/Symfony/Bundle/Resources/config/doctrine_orm.xml]: I just move to follow priority

**Comment by jewome62** [comment]
| Step | Time before | Time After
| ------------- | ------------- | ---
| After Cache clear     |  6521ms | 1539ms
| Hit 1  | 6437ms | 1066ms
| Hit 2  | 6413ms | 1267ms
| Hit 3  | 5555ms | 1176ms
| Hit 4  | 5389ms | 999ms
| Hit 5  | 5390ms | 1182ms

SQL Query Before
 ```sql
SELECT 
  DISTINCT s0_.id AS id_0 
FROM 
  shooting s0_ 
[17 LEFT JOIN]
WHERE 
  s0_.id IN (
    SELECT 
      s18_.id 
    FROM 
      shooting s18_ 
    [17 LEFT JOIN]
    WHERE 
      s18_.agency_id IN (?) 
      OR s18_.id IN (?) 
 ) 
ORDER BY 
  s0_.id ASC 
LIMIT 
  30
```
SQL Query After
```sql
SELECT 
  DISTINCT s0_.id AS id_0 
FROM 
  shooting s0_ 
[17 LEFT JOIN]
WHERE 
  s0_.id IN (
    SELECT 
      s18_.id 
    FROM 
      shooting s18_ 
    [17 LEFT JOIN]
    WHERE 
      s18_.agency_id IN (?) 
      OR s18_.id IN (?) 
    ORDER BY 
      s18_.id ASC
  ) 
LIMIT 30
```

**Comment by jewome62** [comment]
| Step | Time before | Time After | Time fixed version
| ------------- | ------------- | ------------- | ---
| After Cache clear     |  6521ms | 1539ms | 2755ms 
| Hit 1  | 6437ms | 1066ms | 1705ms 
| Hit 2  | 6413ms | 1267ms | 1529ms 
| Hit 3  | 5555ms | 1176ms | 1548ms 
| Hit 4  | 5389ms | 999ms | 1535ms 
| Hit 5  | 5390ms | 1182ms | 1561ms

**Comment by teohhanhui** [comment]
This seems incorrect to me. You'd change the order of results in some cases.

**Comment by teohhanhui** [comment]
The real issue I see here is:

```
[17 LEFT JOIN]
```

How is it a good idea to do such a complex query anyway?

**Comment by jewome62** [comment]
> This seems incorrect to me. You'd change the order of results in some cases.

which case?

**Comment by soyuka** [comment]
Let's close this, sadly there are no perfect solution with query generation.

**Comment by teohhanhui** [comment]
@soyuka Can't we do what's suggested in https://github.com/api-platform/core/pull/2993#issuecomment-522557450?

**Comment by soyuka** [comment]
Not sure to get what you mean. Anyway when queries like these arise we should just write them in plain SQL, remove that useless distinct and do optimizations by improving the model no?

**Comment by teohhanhui** [comment]
See the 2 different versions in these 2 comments:

Unsafe: https://github.com/api-platform/core/pull/2993#issuecomment-522550904
Safe: https://github.com/api-platform/core/pull/2993#issuecomment-522557450

**Comment by soyuka** [comment]
Indeed me neither let's do this then.

File: tests/Bridge/Doctrine/Orm/Extension/FilterEagerLoadingExtensionTest.php [hunk 2]
```diff
@@ -327,9 +327,9 @@ public function testGroupBy()
   SELECT o_2 FROM ApiPlatform\Core\Tests\Fixtures\TestBundle\Entity\DummyCar o_2
   LEFT JOIN o_2.colors colors_2
   WHERE o_2.colors = :foo
+  ORDER BY o_2.colors ASC
 )
 GROUP BY o.colors HAVING counter > 3
-ORDER BY o.colors ASC
```
Comment (soyuka) [inline tests/Bridge/Doctrine/Orm/Extension/FilterEagerLoadingExtensionTest.php]: this needs to be fixed, see last comment

**Pull request closed without merging**